### PR TITLE
AMM-CDA-1: Switch to AmmCdaHybrid enum field

### PR DESCRIPTION
**Pull request by Chralt98**
### What does it do?

Instead of order book and amm separate scoring rules, we use a shared enum field for the hybrid approach.

This also includes the resulting storage migration.

Migration was successful on ``try-runtime` for battery station:

```bash
2024-03-04 16:31:49.940  INFO main zrml_market_commons::migrations: All 755 markets could successfully be decoded.    
2024-03-04 16:31:49.948  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Starting...    
2024-03-04 16:31:50.040  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Upgraded 755 markets.    
2024-03-04 16:31:50.041  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Done!    
2024-03-04 16:31:50.215  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Market counter post-upgrade is 755! 
```

And main-net:

```bash
2024-03-04 16:38:56.388  INFO main zrml_market_commons::migrations: All 360 markets could successfully be decoded.    
2024-03-04 16:38:56.391  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Starting...    
2024-03-04 16:38:56.436  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Upgraded 360 markets.    
2024-03-04 16:38:56.436  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Done!    
2024-03-04 16:38:56.517  INFO main zrml_market_commons::migrations: MigrateScoringRuleAmmCdaHybrid: Market counter post-upgrade is 360!
```

### CODE REVIEW

File: zrml/parimutuel/src/tests/buy.rs [hunk 1]
```diff
@@ -101,8 +101,7 @@ fn buy_fails_if_asset_not_parimutuel_share() {
     });
 }
 
-#[test_case(ScoringRule::Orderbook; "orderbook")]
-#[test_case(ScoringRule::Lmsr; "lmsr")]
+#[test_case(ScoringRule::AmmCdaHybrid; "lmsr")]
```
Comment (maltekliemann) [inline zrml/parimutuel/src/tests/buy.rs:104]:
```suggestion
#[test_case(ScoringRule::AmmCdaHybrid)]
```

File: zrml/parimutuel/src/tests/claim.rs [hunk 2]
```diff
@@ -191,8 +191,7 @@ fn claim_rewards_fails_if_not_resolved(status: MarketStatus) {
     });
 }
 
-#[test_case(ScoringRule::Orderbook; "orderbook")]
-#[test_case(ScoringRule::Lmsr; "lmsr")]
+#[test_case(ScoringRule::AmmCdaHybrid; "lmsr")]
```
Comment (maltekliemann) [inline zrml/parimutuel/src/tests/claim.rs:194]:
```suggestion
#[test_case(ScoringRule::AmmCdaHybrid)]
```

File: zrml/parimutuel/src/tests/refund.rs [hunk 3]
```diff
@@ -65,8 +65,7 @@ fn refund_fails_if_market_not_resolved(status: MarketStatus) {
     });
 }
 
-#[test_case(ScoringRule::Orderbook; "orderbook")]
-#[test_case(ScoringRule::Lmsr; "lmsr")]
+#[test_case(ScoringRule::AmmCdaHybrid; "lmsr")]
```
Comment (maltekliemann) [inline zrml/parimutuel/src/tests/refund.rs:68]:
```suggestion
#[test_case(ScoringRule::AmmCdaHybrid)]
```

**Review by maltekliemann — COMMENTED** [review]
PR description is a bit misleading, as this does actually not contain a storage migration.

File: primitives/src/market.rs [hunk 4]
```diff
@@ -285,8 +285,8 @@ pub struct Deadlines<BN> {
 
 #[derive(TypeInfo, Clone, Copy, Encode, Eq, Decode, MaxEncodedLen, PartialEq, RuntimeDebug)]
 pub enum ScoringRule {
-    Lmsr,
-    Orderbook,
+    // TODO: Do migration: migrate from LMSR, Order Book enum fields to AMM-CDA hybrid field
```
Comment (Chralt98) [inline primitives/src/market.rs:288]:
```suggestion
```

File: zrml/market-commons/src/migrations.rs [hunk 5]
```diff
@@ -15,3 +15,326 @@
...
+                deadlines,
+                scoring_rule: new_scoring_rule,
+                status,
+                report,
+                resolved_outcome,
+                dispute_mechanism,
+                bonds,
+                early_close,
+            });
+            // Don't set up chain to signal that storage is already up to date.
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs]: It was just a copied migration. So I removed the comment c0a7244b0df855f838822549bb0a9d7d62bf7f2c

File: zrml/market-commons/src/migrations.rs [hunk 6]
```diff
@@ -15,3 +15,326 @@
...
+            );
+        });
+    }
+
+    #[test_case(OldScoringRule::Orderbook, Some(ScoringRule::AmmCdaHybrid))]
+    #[test_case(OldScoringRule::Lmsr, Some(ScoringRule::AmmCdaHybrid))]
+    #[test_case(OldScoringRule::Parimutuel, Some(ScoringRule::Parimutuel))]
+    fn on_runtime_upgrade_works_as_expected(
+        old_data: OldScoringRule,
+        new_data: Option<ScoringRule>,
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs]: No reason. It was a copied migration. c0a7244b0df855f838822549bb0a9d7d62bf7f2c

File: zrml/market-commons/src/migrations.rs [hunk 7]
```diff
@@ -15,3 +15,326 @@
...
+const MARKET_COMMONS_REQUIRED_STORAGE_VERSION: u16 = 10;
+const MARKET_COMMONS_NEXT_STORAGE_VERSION: u16 = 11;
+
+#[frame_support::storage_alias]
+pub(crate) type Markets<T: Config> =
+    StorageMap<MarketCommons<T>, Blake2_128Concat, MarketIdOf<T>, OldMarketOf<T>>;
+
+pub struct MigrateScoringRuleAmmCdaHybrid<T>(PhantomData<T>);
+
+/// Deletes all Rikiddo markets from storage and migrates CPMM markets to LMSR.
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/market-commons/src/migrations.rs]: c0a7244b0df855f838822549bb0a9d7d62bf7f2c

**Review by maltekliemann — APPROVED** [review]

**Pull request merged**